### PR TITLE
feat(config): add ${this.mode} template variable to action spec context

### PR DESCRIPTION
**Pull request by edvald**
This allows much easier and cleaner templating around whether an action is in sync or local modes.

Also refactored such that action modes are attached to Action instances as part of config graph resolution, as opposed to dragging along `syncModeDeployNames` etc. arrays.

Other actions can also reference the mode of an action via `${action.<kind>.<name>.mode}`.

### CODE REVIEW

File: core/src/plugins/kubernetes/helm/common.ts [hunk 1]
```diff
@@ -276,32 +271,17 @@ export async function getChartPath(action: Resolved<HelmDeployAction>) {
...
+export async function getValueArgs({ action, valuesPath }: { action: Resolved<HelmDeployAction>; valuesPath: string }) {
   // The garden-values.yml file (which is created from the `values` field in the module config) takes precedence,
   // so it's added to the end of the list.
   const valueFiles = action
     .getSpec("valueFiles")
     .map((f) => resolve(action.getBuildPath(), f))
     .concat([valuesPath])
 
-  const args = flatten(valueFiles.map((f) => ["--values", f]))
+  const args = flatten(valueFiles.map((f) => ["--values", f])).concat()
```
Comment (thsig) [inline core/src/plugins/kubernetes/helm/common.ts:282]: Remove `.concat()`.

File: core/src/garden.ts [hunk 2]
```diff
@@ -979,15 +988,20 @@ export class Garden {
...
           graph,
           config,
           router,
           log: graphLog,
           configsByKey: actionConfigs,
+          mode: actionModes[key] || "default",
         })
+
         graph.addAction(action)
+        // actionConfigs[key] = config
```
Comment (thsig) [inline core/src/garden.ts:1004]: Delete comment.

File: core/test/unit/src/commands/helpers.ts [hunk 3]
```diff
@@ -5,33 +5,3 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/.
  */
-
```
Comment (thsig) [inline core/test/unit/src/commands/helpers.ts:8]: Let's just delete this file.

File: core/test/unit/src/vcs/git.ts [hunk 4]
```diff
@@ -897,6 +897,7 @@ describe("GitHandler", () => {
           error = e
         }
         expect(error).to.be.instanceOf(Error)
+        console.log(error)
```
Comment (thsig) [inline core/test/unit/src/vcs/git.ts:900]: Delete `console.log`.

**Review by thsig — CHANGES_REQUESTED** [review]
Great stuff! This is a lot cleaner than the old way.

Only had a few cleanup-related comments in there. There are also a bunch of newlines added in object literals (didn't want to leave ~20 comments about those), maybe we'd want to clean those up as well.

File: core/src/commands/build.ts [hunk 5]
```diff
@@ -112,8 +112,8 @@ export class BuildCommand extends Command<Args, Opts> {
             action,
             force: opts.force,
             forceActions: [],
-            syncModeDeployNames: [],
-            localModeDeployNames: [],
+
+
```
[the remark this answers is not in the document]
Comment (edvald) [reply core/src/commands/build.ts:116]: I can do this in some scenarios but I _really_ don't want to chase all these down. We can clean that up gradually as we happen upon it.

File: core/test/unit/src/vcs/git.ts (hunk 4 above)
Comment (Walther) [reply core/test/unit/src/vcs/git.ts]: PR made https://github.com/garden-io/garden/pull/3938

File: core/src/commands/build.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (thsig) [reply core/src/commands/build.ts]: Yeah, good point.

**Comment by edvald** [comment]
@thsig should be good to go now.

**Review by thsig — APPROVED** [review]

**Pull request merged**